subsys/usb/class/hid/core.c

```diff
@@ -10,15 +10,64 @@
 #define SYS_LOG_DOMAIN "usb/hid"
 #include <logging/sys_log.h>
 
+#include <misc/byteorder.h>
 #include <usb_device.h>
 #include <usb_common.h>
 
 #include <usb_descriptor.h>
 #include <class/usb_hid.h>
 
-#ifdef CONFIG_USB_COMPOSITE_DEVICE
-#include <composite.h>
-#endif
+struct usb_hid_config {
+	struct usb_if_descriptor if0;
+	struct usb_hid_descriptor if0_hid;
+	struct usb_ep_descriptor if0_int_ep;
+} __packed;
+
+USBD_CLASS_DESCR_DEFINE(primary) struct usb_hid_config hid_cfg = {
+	/* Interface descriptor */
+	.if0 = {
+		.bLength = sizeof(struct usb_if_descriptor),
+		.bDescriptorType = USB_INTERFACE_DESC,
+		.bInterfaceNumber = 0,
+		.bAlternateSetting = 0,
+		.bNumEndpoints = 1,
+		.bInterfaceClass = HID_CLASS,
+		.bInterfaceSubClass = 0,
+		.bInterfaceProtocol = 0,
+		.iInterface = 0,
+	},
+	.if0_hid = {
+		.bLength = sizeof(struct usb_hid_descriptor),
+		.bDescriptorType = USB_HID_DESC,
+		.bcdHID = sys_cpu_to_le16(USB_1_1),
+		.bCountryCode = 0,
+		.bNumDescriptors = 1,
+		.subdesc[0] = {
+			.bDescriptorType = USB_HID_REPORT_DESC,
+			/*
+			 * descriptor length needs to be set
+			 * after initialization
+			 */
+			.wDescriptorLength = 0,
+		},
+	},
+	.if0_int_ep = {
+		.bLength = sizeof(struct usb_ep_descriptor),
+		.bDescriptorType = USB_ENDPOINT_DESC,
+		.bEndpointAddress = CONFIG_HID_INT_EP_ADDR,
+		.bmAttributes = USB_DC_EP_INTERRUPT,
+		.wMaxPacketSize =
+			sys_cpu_to_le16(
+			CONFIG_HID_INTERRUPT_EP_MPS),
+		.bInterval = 0x09,
+	},
+};
+
+static void usb_set_hid_report_size(u16_t report_desc_size)
+{
+	hid_cfg.if0_hid.subdesc[0].wDescriptorLength =
```

Comment:
@jfischer-phytec-iot 
Johann, this assignment was causing an unaligned store exception when I ran the rebased code on Intel S1000. I think this would happen on other architectures that require aligned accesses.
`wDescriptorLength` is a `u16_t`.
The descriptors are arranged in memory such that
The device descriptor is first and occupies 18 bytes
The configuration descriptor is next and occupies 9 bytes
The interface descriptor occupies 9 bytes
The HID class descriptor starts at a relative offset of 18+9+9 = 36 bytes
The `usb_hid_class_subdescriptor` starts at an offset of 36 + 6 = 42 bytes
Within `usb_hid_class_subdescriptor`, the first field is a 1 byte `bDescriptorType` and the 2nd field is a 2 byte `wDescriptorLength` and this 2 byte field starts at an odd address/offset.
And this causes an unaligned data store exception.
I had to use byte assignments temporarily to get around this exception, and I wanted to bring this to your attention to see if you expected this when you reworked the code.

Reply:
@sathishkuttan Why should not it happen before this PR?

can you please test this:
```
--- a/subsys/usb/class/hid/core.c
+++ b/subsys/usb/class/hid/core.c
@@ -68,8 +68,8 @@ USBD_CLASS_DESCR_DEFINE(primary) struct usb_hid_config hid_cfg = {
 
 static void usb_set_hid_report_size(u16_t report_desc_size)
 {
-       hid_cfg.if0_hid.subdesc[0].wDescriptorLength =
-               sys_cpu_to_le16(report_desc_size);
+       UNALIGNED_PUT(sys_cpu_to_le16(report_desc_size),
+                     &hid_cfg.if0_hid.subdesc[0].wDescriptorLength);
 }
 
 static struct hid_device_info {
```
I suspected it will be necessary at several places.

Reply:
@jfischer-phytec-iot 
Good question. I just tested with the 1.12 release tag and surprisingly, the wDescriptorLength is also at an odd address but there is no exception!
Looks like something else changed in another commit and this is not due to your commit in this PR.